firedrake/extrusion_utils.py

```diff
@@ -43,119 +46,156 @@ def make_extruded_coords(extruded_topology, base_coords, ext_coords,
                                and vert_space.family() in ['Lagrange',
                                                            'Discontinuous Lagrange']):
         raise RuntimeError('Extrusion of coordinates is only possible for a P1 or P1dg interval unless a custom kernel is provided')
+    layer_height = op2.Global(1, layer_height, dtype=RealType)
     if kernel is not None:
-        pass
-    elif extrusion_type == 'uniform':
-        kernel = op2.Kernel("""
-static inline void pyop2_kernel_uniform_extrusion(double *ext_coords,
-                                                  const double *base_coords,
-                                                  const double *layer_height,
-                                                  int layer) {
-    for ( int d = 0; d < %(base_map_arity)d; d++ ) {
-        for ( int c = 0; c < %(base_coord_dim)d; c++ ) {
-            ext_coords[2*d*(%(base_coord_dim)d+1)+c] = base_coords[d*%(base_coord_dim)d+c];
-            ext_coords[(2*d+1)*(%(base_coord_dim)d+1)+c] = base_coords[d*%(base_coord_dim)d+c];
-        }
-        ext_coords[2*d*(%(base_coord_dim)d+1)+%(base_coord_dim)d] = *layer_height * (layer);
-        ext_coords[(2*d+1)*(%(base_coord_dim)d+1)+%(base_coord_dim)d] = *layer_height * (layer + 1);
-    }
-}""" % {'base_map_arity': base_coords.cell_node_map().arity, 'base_coord_dim': base_coords.function_space().value_size},
-            "pyop2_kernel_uniform_extrusion")
+        op2.ParLoop(kernel,
+                    ext_coords.cell_set,
+                    ext_coords.dat(op2.WRITE, ext_coords.cell_node_map()),
+                    base_coords.dat(op2.READ, base_coords.cell_node_map()),
+                    layer_height(op2.READ),
+                    pass_layer_arg=True,
+                    is_loopy_kernel=True).compute()
+        return
+    ext_fe = create_element(ext_coords.ufl_element())
+    ext_shape = ext_fe.index_shape
+    base_fe = create_element(base_coords.ufl_element())
+    base_shape = base_fe.index_shape
+    data = []
+    data.append(lp.GlobalArg("ext_coords", dtype=ScalarType, shape=ext_shape))
+    data.append(lp.GlobalArg("base_coords", dtype=ScalarType, shape=base_shape))
+    data.append(lp.GlobalArg("layer_height", dtype=ScalarType, shape=()))
```

Comment:
Above in line 49 layer_height was declared as RealType, here it is passed as ScalarType. Is that for a reason?

Reply:
Fixed.